File: 494-Target Sum.py
# hash
# Runtime: 228 ms, faster than 78.46% of Python3 online submissions for Target Sum.
# Memory Usage: 12.6 MB, less than 100.00% of Python3 online submissions for Target Sum.
class Solution:
    def findTargetSumWays(self, nums: 'List[int]', S: 'int') -> 'int':
        dict = {0: 1}
        for item in nums:
            tmp_dict = {}
            for tmp_sum in dict:
                tmp_dict[tmp_sum + item] = tmp_dict.get(tmp_sum + item, 0) + dict[tmp_sum]
                tmp_dict[tmp_sum - item] = tmp_dict.get(tmp_sum - item, 0) + dict[tmp_sum]
            dict = tmp_dict
        return dict.get(S, 0)

File: Algorithms/C++/494-Target_Sum.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Target Sum.
// Memory Usage: 8.5 MB, less than 76.92% of C++ online submissions for Target Sum.
class Solution {
public:
    int findTargetSumWays(vector<int>& nums, int S) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum < S || (sum + S) & 1)
            return 0;
        // dp
        S = (S + sum) >> 1;
        vector<int> dp(S + 1, 0);
        dp[0] = 1;
        for (int n : nums)
            for (int i = S; i >= n; --i)
                dp[i] += dp[i - n];
        return dp[S];
    }
};
